abstract_domains/lattice.py

```diff
@@ -1,93 +1,74 @@
 from abc import ABC, abstractmethod
-from enum import Enum
 from functools import reduce
 from typing import List
 
 
 class Lattice(ABC):
-    class Kind(Enum):
-        """Kind of a lattice element."""
-        Bottom = -1  # bottom element
-        Default = 0
-        Top = 1  # top element
-
-    class Internal(object):
-        def __init__(self, kind: 'Lattice.Kind'):
-            """Lattice internal representation.
-
-            :param kind: kind of lattice element
-            """
-            self._kind = kind
-
-        @property
-        def kind(self):
-            return self._kind
-
-        @kind.setter
-        def kind(self, kind: 'Lattice.Kind'):
-            self._kind = kind
-
-        def bottom(self):
-            """Return the bottom lattice element internal representation.
-            
-            :return: current internal representation modified to be the bottom lattice element
-            """
-            self.kind = Lattice.Kind.Bottom
-            return self
+    def __init__(self):
+        """Creates a new lattice element and sets it to the default.
+        """
+        # initialize with default
+        self.default()
 
-        def top(self):
-            """Return the top lattice element internal representation.
-            
-            :return: current internal representation modified to be the top lattice element
-            """
-            self.kind = Lattice.Kind.Top
-            return self
+    def __eq__(self, other: 'Lattice'):
+        if isinstance(other, self.__class__):
+            return repr(self) == repr(other)
+        return False
 
-    def __init__(self, kind: Kind = Kind.Default):
-        """Lattice representation.
-        Account for lattice operations by modifying the current internal representation.
+    def __hash__(self):
+        return hash(repr(self))
 
-        :param kind: kind of lattice element
+    def __ne__(self, other: 'Lattice'):
+        return not (self == other)
+
+    @abstractmethod
+    def __repr__(self):
+        """Unambiguous string representing the current lattice element.
+
+        :return: unambiguous representation string
         """
-        self._internal = Lattice.Internal(kind)
 
-    @property
-    def internal(self):
-        return self._internal
+    def __str__(self):
+        """Human-readable string representing this internal.
 
-    @property
-    def kind(self):
-        return self.internal.kind
+        :return: human-readable string
+        """
+        return repr(self)
```

Comment:
We can get rid of the __str__ method if it just calls the __repr__ method since I think it does that by default.